### PR TITLE
Fix CloseScope0Tests hang

### PR DESCRIPTION
**Pull request by EricYangIBM**
Fixes: #13158
Signed-off-by: Eric Yang <eric.yang@ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
Test via grinder https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/135/

**Comment by EricYangIBM** [comment]
```java
Thread t1 = new Thread(()->{
	try {
		synchronized (synch1) {
			synch1.wait();
		}
		close.invoke(scope);
	} catch (InvocationTargetException e) {
		// This is the expected behaviour (throws IllegalStateException)
		expected = e.getCause();
	} catch (InterruptedException | IllegalAccessException e) {
		e.printStackTrace();
	} finally {
		synchronized (synch2) {
			synch2.notifyAll();
		}
	}
}, "ScopeCloserThread");

class MyRunnable implements Runnable {
	public void run() {
		try {
			synchronized (synch1) {
				synch1.notify();
			}
			synchronized (synch2) {
				synch2.wait();
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}

MyRunnable r = new MyRunnable();
Thread t2 = new Thread(()->{
	try {
		runInScoped.invoke(null, r, scope);
	} catch (Exception e) {
		e.printStackTrace();
	}
}, "RunInScopeThread");

t1.start();
t2.start();

synchronized (synch2) {
	synch2.wait();
}
```
@tajila I think the code may get stuck in three scenarios:
1. t2 goes first and notifies before t1 waits
2. t1 notifyAlls before t2 synch2.wait
3. t1 notifyAlls before the main thread synch2.wait

I added timeouts on the waits but I don't think this is a good solution. The t2 synch2.wait has a longer timeout because for scenario 1 we don't want t2 to exit before t1 closes the scope.

**Comment by tajila** [comment]
you can try the following.

in main thread
```
t1.start();

while (!t1Started) {
     Thread.yield()
}

t2.start();
```

in T1

```
t1Started = true

synchronized (synch1) {
	synch1.wait();
}
close.invoke(scope);
```

Essentially, try to force an ordeering constraint between when t1 and t2 starts

Also `t1Started` needs to be `volatile`

**Comment by pshipton** [comment]
Grinder results, didn't hang but didn't pass either.
https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/137/tapResults/
```
FAILED: closeScopeDuringAccess
org.testng.TestException: 
Method TestCloseScope0.closeScopeDuringAccess()[pri:0, instance:org.openj9.test.foreignMemoryAccess.TestCloseScope0@83f4b58a] should have thrown an exception of type class java.lang.IllegalStateException
	at org.testng.internal.ExpectedExceptionsHolder.noException(ExpectedExceptionsHolder.java:80)
	at org.testng.internal.Invoker.considerExceptions(Invoker.java:1139)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:612)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:716)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:988)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.TestNG.privateMain(TestNG.java:1354)
	at org.testng.TestNG.main(TestNG.java:1323)
```

**Comment by tajila** [comment]
Re-running grinder with latest changes

https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/138/

**Comment by EricYangIBM** [comment]
Should I put a yield in t2 as well? (e.g. before the `synchronized (synch1)`)
Also will yield guarantee the ordering? looking at the documentation: 
> A hint to the scheduler that the current thread is willing to yield its current use of a processor. The scheduler is free to ignore this hint.

**Comment by tajila** [comment]
The  yield is not there to guarantee ordering, its there to not waste CPU while waiting for t1 to start

**Comment by tajila** [comment]
ordering is guaranteed by the `t1Started` flag

File: test/functional/Java16andUp/src/org/openj9/test/foreignMemoryAccess/TestCloseScope0.java [hunk 1]
```diff
@@ -111,6 +112,8 @@ public static void closeScopeDuringAccess() throws Throwable {
 
 		Thread t1 = new Thread(()->{
 			try {
+				t1Started = true;
```
Comment (tajila) [inline test/functional/Java16andUp/src/org/openj9/test/foreignMemoryAccess/TestCloseScope0.java:115]: To be safe `t1Started = true;` should be moved in the `synchronized (synch1) {` block

**Comment by tajila** [comment]
The test passed, but I have one minor comment

**Comment by EricYangIBM** [comment]
Is it possible that t1's notifyAll happens before main thread's and/or t2's wait?

**Comment by tajila** [comment]
> Is it possible that t1's notifyAll happens before main thread's and/or t2's wait?

Yes, I think you can add a flag in t2

```
synchronized (synch1) {
		synch1.notify();
	}
	synchronized (synch2) {
                t2Waiting = true
		synch2.wait();
	}
```
and in t1 spin on that flag before calling notifyAll

The main thread doesn't need the wait, it can just join on t1 and t2

**Comment by EricYangIBM** [comment]
Done. I also moved the while (!t1Started) loop into t2 for consistency, is this ok?

**Comment by tajila** [comment]
> Done. I also moved the while (!t1Started) loop into t2 for consistency, is this ok?

ya thats fine

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
Grinder https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/140/

**Comment by tajila** [comment]
Jenkins test sanity,extended plinux jdk16

**Pull request merged**